remoteappmanager/cli/remoteapprest/__main__.py

```diff
@@ -0,0 +1,209 @@
+
+
+class Credentials:
+    """Data class to hold the credentials extracted from the credential file.
+    """
+    def __init__(self, url, username, cookies):
+        self.url = url
+        self.username = username
+        self.cookies = cookies
+
+    @classmethod
+    def from_file(cls, credentials_file):
+        """Extracts the authorization info from the credentials file.
+        Returns a tuple with url, username, and a dict of credentials
+        cookies"""
+        with open(credentials_file, "r") as f:
+            lines = f.readlines()
+
+        url = lines[0].strip()
+        username = lines[1].strip()
+        cookies = {}
+        for line in lines[2:]:
+            k, v = line.split('=', 1)
+            cookies[k.strip()] = v.strip()
+
+        return cls(url, username, cookies)
+
+    def write(self, credentials_file):
+        """Stores the credentials in a credentials file."""
+        with open(credentials_file, "w") as f:
+            f.write("{}\n".format(self.url))
+            f.write("{}\n".format(self.username))
+            for k, v in self.cookies.items():
+                f.write("{}={}\n".format(k, v))
+
+
+class RemoteAppRestContext:
+    """The click context passed around."""
+    credentials_file = None
+    credentials = None
+
+
+@click.group()
+@click.option("--credentials-file",
+              default=os.path.expanduser("~/.remoteapprest"),
+              help="Specify a different credentials file.")
+@click.pass_context
+def cli(ctx, credentials_file):
+    """Command line interface to start, stop, and inquire applications
+    on the remote application server."""
+    ctx.obj.credentials_file = credentials_file
+    try:
+        ctx.obj.credentials = Credentials.from_file(credentials_file)
+    except IOError:
+        ctx.obj.credentials = None
+
+
+@cli.command()
+@click.argument("url")
+@click.option("--username", prompt=True)
+@click.option('--password',
+              prompt=True,
+              confirmation_prompt=False,
+              hide_input=True)
+@click.pass_context
+def login(ctx, url, username, password):
+    """Performs login on the remote server at the specified URL."""
+    login_url = urljoin(url, "/hub/login")
+
+    payload = {"username": username, "password": password}
+
+    # Unfortunately, jupyterhub handles the afterlogin with an immediate
+    # redirection, meaning that we have to check for a 302 and prevent
+    # redirection in order to capture the cookies.
+    try:
+        response = requests.post(login_url, payload, verify=False,
+                                 allow_redirects=False)
+    except Exception as e:
+        print("Could not perform request. {}".format(e), file=sys.stderr)
+        sys.exit(1)
+
+    if response.status_code == 302:
+        cookies_dict = requests.utils.dict_from_cookiejar(response.cookies)
+        cred = Credentials(url, username, cookies_dict)
+        cred.write(ctx.obj.credentials_file)
```

Comment:
But line 67 assigns it to None if IOError occurs...I suppose then you get IOError again when you try to write to it...